### PR TITLE
mqueue: refactor of mqueue

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary

mqueue: refactor of mqueue

    1. mainly for long time within critical_section
    2. move out memcpy buffer from enter_critical_section()
    3. let mq_send() return fail when mq buffer full in ISR

after these optimize:
the mq_timedsend()/mq_timedrecv() critical time, which use critmon tools to measure:
from 100us -> 60us

## Impact

mqueue

## Testing

bes board & sim

### CODE REVIEW

**Comment by GUIDINGLI** [comment]
This should submit after:
https://github.com/apache/nuttx/pull/13410

**Comment by lucasssvaz** [comment]
@GUIDINGLI I don't work in the Nuttx project anymore. Maybe @tmedicci can help ?

File: sched/mqueue/mq_rcvinternal.c [hunk 1]
```diff
@@ -156,139 +147,90 @@ int nxmq_wait_receive(FAR struct mqueue_inode_s *msgq,
     }
 #endif
 
+  if (abstime)
```
Comment (pkarashchenko) [inline sched/mqueue/mq_rcvinternal.c:150]: Do we need to check `abstime->tv_sec` and `abstime->tv_nsec` are non zero or non-NUL `abstime` ensures that non zero timeout is passed here?

Comment (GUIDINGLI) [reply sched/mqueue/mq_rcvinternal.c:150]: already check in file_mq_timedsend_internal() and file_mq_timedreceive_internal()

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by fdcavalcanti** [comment]
Hi folks.
Just to let you know, this PR broke WiFi tests on all Espressif SoCs.
`ifconfig` is behaving differently, not changing the interface state after a reboot.

**Comment by xiaoxiang781216** [comment]
@GUIDINGLI please take a look.

**Comment by Javinator9889** [comment]
I have the feeling this is breaking the mqueues under certain conditions - for example, heavy load. We were using NuttX 12.1.0 and we've updated to 12.9.0. We've reached a situation where:

- A thread is blocked waiting for `MQNOTFULL`.
- Another thread is blocked waiting for `MQNOTEMPTY`.

Both of them remain locked, and we're reaching a deadlock situation in there :/

IMO these refactors should be driven with special care. Finding a bug among +1300 changes is going to be messy 😅

**Comment by xiaoxiang781216** [comment]
@GUIDINGLI could you look at this problem.

**Comment by GUIDINGLI** [comment]
> I have the feeling this is breaking the mqueues under certain conditions - for example, heavy load. We were using NuttX 12.1.0 and we've updated to 12.9.0. We've reached a situation where:
> 
> * A thread is blocked waiting for `MQNOTFULL`.
> * Another thread is blocked waiting for `MQNOTEMPTY`.
> 
> Both of them remain locked, and we're reaching a deadlock situation in there :/
> 
> IMO these refactors should be driven with special care. Finding a bug among +1300 changes is going to be messy 😅

Can you provide detail backtrace for the deadlock？

**Comment by Javinator9889** [comment]
I don't have a detailed backtrace of the deadlock - sorry. I was checking with `ps` on NSH and found that the producer thread was stuck on `MQNOTFULL` and the consumer was on `MQNOTEMPTY`

**Comment by GUIDINGLI** [comment]
> > I have the feeling this is breaking the mqueues under certain conditions - for example, heavy load. We were using NuttX 12.1.0 and we've updated to 12.9.0. We've reached a situation where:
> > 
> > * A thread is blocked waiting for `MQNOTFULL`.
> > * Another thread is blocked waiting for `MQNOTEMPTY`.
> > 
> > Both of them remain locked, and we're reaching a deadlock situation in there :/
> > IMO these refactors should be driven with special care. Finding a bug among +1300 changes is going to be messy 😅
> 
> Can you provide detail backtrace for the deadlock？

Also，please provide the testcase，then I can reproduce it

**Comment by Javinator9889** [comment]
> > > I have the feeling this is breaking the mqueues under certain conditions - for example, heavy load. We were using NuttX 12.1.0 and we've updated to 12.9.0. We've reached a situation where:
> > > 
> > > * A thread is blocked waiting for `MQNOTFULL`.
> > > * Another thread is blocked waiting for `MQNOTEMPTY`.
> > > 
> > > Both of them remain locked, and we're reaching a deadlock situation in there :/
> > > IMO these refactors should be driven with special care. Finding a bug among +1300 changes is going to be messy 😅
> > 
> > 
> > Can you provide detail backtrace for the deadlock？
> 
> Also，please provide the testcase，then I can reproduce it

There are some environment conditions that seem to be affecting. I cannot give you a test case to easily reproduce it, sorry. I'll continue my investigation

**Pull request merged**